### PR TITLE
Fix errors when renaming/moving/deleting global scripts

### LINKED ISSUE #67056: GDScript 2.0: Class YYY hides a global script class: Caused by "orphan" script in .tscn due to renamed files
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0.beta2

### System information

Windows 10

### Issue description

I "randomly" had `class YYY hides a global script class` error, and it seems to be because .tscn files hold a copy of the attached script. Thus, if we rename or move that script (for example capitalisation which is poorly handled on windows systems) when end with a duplicated script.
If said script has a class_name, we end up with duplicated class.

Should scenes be able to have a script not stored in a .gd ? This error is really counterintuitive to find, as you'd be tempted to use find feature which do not list .tscn content.

### Steps to reproduce

Create a script with a class_name.
Attach it to a node.
Rename that script with an uppercase on windows.
Try to run your game, the error should pop up.

### Minimal reproduction project

_No response_

### LINKED ISSUE #81867: "FIle not found" errors after deleting folder containing GDScript file with custom class_name
### Godot version

4.1.1

### System information

Windows 11

### Issue description

When a GDScript file with a custom class_name is deleted, the class is also removed from the editor.  However, if the folder is deleted (which contained the same GDScript file) the class is seemingly not removed from the editor and when you search for nodes, it tries to open the missing file and reports errors.

### Steps to reproduce

1. Create a new folder in the resource tree (eg example).
2. Inside this new folder, create a new GDScript file to Extend Node3D (although can be anything, I think)
3. Inside the new GDScript file, add the following line under the Extends Node3D line:
    class_name someclass
4. Save the GDScript file as myscript.gd.
5. Create a new Scene, select Other Node as the Root Node and search for your new class (someclass, in this example)
6. The option to extend your new scene from this class and GDScript file is available.
7. Cancel the Create New Node dialog
8. Delete the folder created in step 1
9. Attempt step 5 again - you will get errors in the console as follows:

Attempt to open script 'res://example/myscript.gd' resulted in error 'File not found'.
  Failed loading resource: res://example/myscript.gd. Make sure resources have been imported by opening the project in the editor at least once.
  editor/create_dialog.cpp:241 - Condition "scr.is_null()" is true.

Closing and reopening Godot has no effect.  The only way to fix this, is to create the same GDScript file again and delete the file rather that the folder it is in.  Deleting the file itself in step 8 results in no errors being reported.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by Maran23**
Fixes: https://github.com/godotengine/godot/issues/67056
Fixes: https://github.com/godotengine/godot/issues/81867
Fixes the root cause in (but there is also some discussion about regnerating the script cache): https://github.com/godotengine/godot/issues/78642

When renaming or moving global scripts, the following errors can appear:
- Attempt to open script 'xxx' resulted in error 'File not found'.
- Failed loading resource: xxx. Make sure resources have been imported by opening the project in the editor at least once.
- Parser Error: Class 'xxx' hides a global script class.

When deleting scripts, errors appear when opening the 'Create Node Dialog' as the script cache still contains the removed global scripts. The following errors can appear:
- Attempt to open script 'xxx' resulted in error 'File not found'.
- Failed loading resource: xxx. Make sure resources have been imported by opening the project in the editor at least once.
editor/create_dialog.cpp:241 - Condition "scr.is_null()" is true.

All this errors can be fixed by correctly handling the cases. They involves removing the old path and adding the new one (if not deleted) to the ScriptServer. This is somewhat similar if the file is moved or deleted outside Godot and detected by the file watcher, but more specialized for this particular usecase, since we know the old and the new path / know what the user just did.

You can verify this in the following Test Project:
[MoveClass.zip](https://github.com/godotengine/godot/files/14859080/MoveClass.zip)

Just move the `CScript.gd` file around and start the game. You will eventually get the error mentioned above without this patch.
You can also delete the folders and verify, that there will not be any error when opening the 'Create Node Dialog'.

### CODE REVIEW

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -1710,6 +1745,13 @@ void FileSystemDock::_make_scene_confirm() {
 }
 
 void FileSystemDock::_resource_removed(const Ref<Resource> &p_resource) {
+	const Script *scr = Object::cast_to<Script>(*p_resource);
+	if (scr != nullptr) {
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:1749]:
```suggestion
	const Ref<Script> &scr = p_resource
	if (scr.is_valid()) {
```

**Review by KoBeWi — APPROVED** [review]

File: editor/filesystem_dock.cpp (hunk 1 above)
Comment (AThousandShips) [reply editor/filesystem_dock.cpp:1749]:
```suggestion
	const Ref<Script> &scr = p_resource;
	if (scr.is_valid()) {
```

File: editor/filesystem_dock.cpp [hunk 2]
```diff
@@ -1561,10 +1561,42 @@ void FileSystemDock::_try_duplicate_item(const FileOrFolder &p_item, const Strin
...
 
 void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, String> &p_renames, const HashMap<String, ResourceUID::ID> &p_uids) const {
-	// Update the paths in ResourceUID, so that UIDs remain valid.
-	for (const KeyValue<String, ResourceUID::ID> &pair : p_uids) {
-		if (p_renames.has(pair.key)) {
-			ResourceUID::get_singleton()->set_id(pair.value, p_renames[pair.key]);
+	for (const KeyValue<String, String> &pair : p_renames) {
+		// Update UID path.
+		String old_path = pair.key;
+		String new_path = pair.value;
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:1567]:
```suggestion
		const String &old_path = pair.key;
		const String &new_path = pair.value;
```

File: editor/filesystem_dock.cpp (hunk 1 above)
Comment (Maran23) [reply editor/filesystem_dock.cpp:1749]: Thanks, good suggestion!

File: editor/filesystem_dock.cpp (hunk 2 above)
Comment (Maran23) [reply editor/filesystem_dock.cpp]: Thanks, makes sense to use the address instead!

File: editor/filesystem_dock.cpp [hunk 3]
```diff
@@ -1561,10 +1561,42 @@ void FileSystemDock::_try_duplicate_item(const FileOrFolder &p_item, const Strin
...
+			ResourceUID::get_singleton()->set_id(I->value, new_path);
+		}
+
+		ScriptServer::remove_global_class_by_path(old_path);
+
+		int index = -1;
+		EditorFileSystemDirectory *efd = EditorFileSystem::get_singleton()->find_file(old_path, &index);
+
+		if (!efd || index < 0) {
+			// The file was removed
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:1580]:
```suggestion
			// The file was removed.
```

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**